src/System.Buffers.Primitives/System/Buffers/Transformation.cs

```diff
@@ -8,6 +8,10 @@ public interface IBufferOperation
     {
         OperationStatus Execute(ReadOnlySpan<byte> input, Span<byte> output, out int consumed, out int written);
     }
+    public interface IBufferTransformation
+    {
+        OperationStatus Transform(Span<byte> buffer, int dataLength, out int written);
```

Comment:
As part of updating the Base64 APIs, I will change the in-place methods to return bool instead of OperationStatus. That won't break your scenario, correct?

Reply:
I still need to know if: a) the buffer was too small, b) the data in the buffer was invalid

Reply:
So, if the in place statics methods returned bool, would the Transform implementation look something like this:
```C#
public override OperationStatus Transform(Span<byte> buffer, int dataLength, out int written)
{
   if (dataLength > buffer.Length) 
      return OperationStatus.DestinationTooSmall;
   if (Utf8ToBytesInPlace(buffer.Slice(0, dataLength), out var consumed, out written)) 
      return OperationStatus.Done;
   return OperationStatus.InvalidData;
}
```

Reply:
But then I have to compute, which would cause the buffer to be walked twice. Why would we want to do it? So we can not return an emum? Well, I take performance APIs that return an enum over APIs that force you to walk the data twice and return a bool.

Reply:
> But then I have to compute, which would cause the buffer to be walked twice

You wouldn't walk the buffer twice since getting the required length is just an O(1) computation based on the length. I meant to use GetMaxEncodedToUtf8Length in the sample above, not ComputeRequiredLength.

Reply:
But you’d have to compute *anyway* if your in-place transformation could possibly increase the amount of data in the buffer. That’s because the operation is necessarily destructive and you have no way to signal partial completion to the caller in a manner that he can resume since the data he would need to pass into you has been overwritten. The only way to avoid this double computation is to require the caller to give you a large enough buffer ahead of time or to require in-place transforms to never produce more output than it consumes, which by definition means you can never receive “output too small” as a status code.

Reply:
It's O(1) only in the case of some transformations

Reply:
Yeah, we have to decide what to do with the destructive nature of some of these transformations. It's a big problem.